src/lib/optimizer/optimizer.hpp

```diff
@@ -32,7 +37,9 @@ class Optimizer final {
    */
   void add_rule(std::unique_ptr<AbstractRule> rule);
 
-  std::shared_ptr<AbstractLQPNode> optimize(std::shared_ptr<AbstractLQPNode> input) const;
+  std::shared_ptr<AbstractLQPNode> optimize(
+      std::shared_ptr<AbstractLQPNode> input,
+      std::shared_ptr<std::vector<OptimizerRuleMetrics>> rule_durations = nullptr) const;
```

Comment:
Why do we need a shared_ptr?

Reply:
references and default parameters seem not to like each other... what's your suggestion here?

Reply:
Ah, makes sense. No good solution yet. I'm not a fan of output parameters, but changing the return type to a more complicated one just to include information that does not pertain to the actual job sounds worse.